OADP 1.2 Data Mover with Ceph (#60936)
openshift/openshift-docs: modules/oadp-ceph-cephfs.adoc

Reviewer: This seems like it is three procedures:
[Steps 1-4] Preparing
[Step 5] Backing up
[Step 6] Restoring

Breaking this into three modules would improve scannability and better adhere to mod docs guidelines.

// Module included in the following assemblies:
//
// * backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.adoc

:_content-type: PROCEDURE

> [id="oadp-ceph-preparing-crs_{context}"]

Reviewer 2: Please can you check this ID as it looks like it is duplicated in oadp-ceph-preparing-crs.adoc

Author: That's one of them...

+
Example DPA CR
+
[source,yaml]
----
apiVersion: oadp.openshift.io/v1alpha1
kind: DataProtectionApplication
metadata:
  name: velero-sample
  namespace: openshift-adp
spec:
  backupLocations:
    - velero:
        config:
          profile: default
          region: us-east-1
        credential:
          key: cloud
          name: cloud-credentials
        default: true
        objectStorage:
          bucket: <my-bucket>
          prefix: velero
       provider: aws
    configuration:
      restic:
        enable: false
      velero:
        defaultPlugins:
          - openshift
          - aws
          - csi
          - vsm
    features:
      dataMover:
        credentialName: <restic-secret-name>
        enable: true
        volumeOptionsForStorageClasses:
          ocs-storagecluster-cephfs:
            sourceVolumeOptions:
              accessMode: ReadOnlyMany
              cacheAccessMode: ReadWriteMany
              cacheStorageClassName: ocs-storagecluster-cephfs
              storageClassName: ocs-storagecluster-cephfs-shallow

> ----

Author: A real improvement -- thanks!

. To back up data:

.. Create a `Backup` CR:
+
Example `Backup` CR
+
[source,yaml]
----
apiVersion: velero.io/v1
kind: Backup
metadata:
  name: <backup-name>
  namespace: <protected-ns>
spec:
  includedNamespaces:

> - <app-ns>

Author: @jeana-redhat Will double-check the original.

> - <app-ns>

Author: @jeana-redhat Good catch! My mistake.

storageLocation: velero-sample-1
----

> .. Monitor the Data Mover backup and its artifacts by running the script named link:https://github.com/openshift/oadp-operator/blob/master/docs/examples/datamover_resources.sh[Volumemover live debug]

Reviewer: In general, we are not to link out to GitHub scripts. Has this been validated by QE?

See [this thread](https://redhat-internal.slack.com/archives/C85JYPHL3/p1686057008204659) for more discussion on this topic, including concerns about scripts on the `main`/`master` branch.
```suggestion
.. Monitor the Data Mover backup and its artifacts by running the script named link:https://github.com/openshift/oadp-operator/blob/master/docs/examples/datamover_resources.sh[Volumemover live debug].
```

Author: @jeana-redhat Good thing you asked -- QE vetted the PR but when i asked about the script specifically, the lead engineer replied that the script is not for downstream. Will remove it when I fix this PR.

.. To check the progress of all the `VolumeSnapshotBackup` CRs, run the following command:
+
[source, terminal]
----
$ oc get vsb -n <app-ns>
----

.. To check the progress of a specific `VolumeSnapshotBackup` CR, run the following command:
+
[source,terminal]
----
$ oc get vsb <vsb-name> -n <app-ns> -ojsonpath="{.status.phase}`
----

> .. Wait several minutes and check if the `VolumeSnapshotBackup` CR has the status `Completed`.

Reviewer: ```suggestion
.. Wait several minutes until the `VolumeSnapshotBackup` CR has the status `Completed`.
```

.. Verify that is at least one snapshot in the object store that is given in the Restic `secret`. You can check for this snapshot in your targeted `BackupStorageLocation` that has a prefix of `/<OADP-namespace>`.

. To restore data:

> .. Verify that the application namespace is deleted, as well as any `VolumeSnapshotContent` CRs that were created during backup.

Reviewer: How do they do this?

> .. Create a `Restore` CR:

Reviewer: ```suggestion
.. Create a `Restore` CR similar to the following:
```

+

> Example `Restore` CR

Reviewer: ```suggestion
.Example `Restore` CR
```

+
[source,yaml]
----
apiVersion: velero.io/v1
kind: Restore
metadata:
  name: <restore-name>
  namespace: <protected-ns>
spec:
  backupName: <previous-backup-name>
----

> .. Monitor the Data Mover backup and its artifacts by running the script named link:https://github.com/openshift/oadp-operator/blob/master/docs/examples/datamover_resources.sh[Volumemover live debug]

Reviewer: Same issue with scripts as above.
```suggestion
.. Monitor the Data Mover backup and its artifacts by running the script named link:https://github.com/openshift/oadp-operator/blob/master/docs/examples/datamover_resources.sh[Volumemover live debug].
```

.. To check the progress of all the `VolumeSnapshotRestore` CRs, run the following command:
+
[source, terminal]
----
$ oc get vsr -n <app-ns>
----

.. To check the progress of a specific `VolumeSnapshotRestore` CR, run the following command:
+
[source,terminal]
----
$ oc get vsr <vsr-name> -n <app-ns> -ojsonpath="{.status.phase}
----

> .. Verify that your application data has been restored:

Reviewer: ```suggestion
.. Verify that your application data has been restored by running the following command:
```

----

.. Wait several minutes until the `VolumeSnapshotBackup` CR has the status `Completed`.
.. Verify that there is at least one snapshot in the object store that is given in the Restic `secret`. You can check for this snapshot in your targeted `BackupStorageLocation` storage provider that has a prefix of `/<OADP-namespace>`.

. To restore data, complete the following steps:

.. Delete the application namespace by running the following command:
+
[source,terminal]
----
$ oc delete vsb -n <app_namespace> --all
----

.. Delete any `VolumeSnapshotContent` CRs that were created during backup by by running the following command:
+
[source,terminal]
----
$ oc oc delete volumesnapshotcontent --all
----

.. Create a `Restore` CR similar to the following:
+
.Example `Restore` CR
+
[source,yaml]
----
apiVersion: velero.io/v1
kind: Restore
metadata:
  name: <restore-name>
  namespace: <protected-ns>
spec:
  backupName: <previous-backup-name>
----

.. Monitor the progress of the `VolumeSnapshotRestore` CRs by doing the following:
... To check the progress of all the `VolumeSnapshotRestore` CRs, run the following command:
+
[source, terminal]
----
$ oc get vsr -n <app-ns>
----

> ... To check the progress of a specific `VolumeSnapshotRestore` CR, run tc:

Reviewer: typo
```suggestion
... To check the progress of a specific `VolumeSnapshotRestore` CR, run the following command:
```

. To back up data, complete the following steps:

.. Create a `Backup` CR similar to the following:
+
.Example `Backup` CR
+
[source,yaml]
----
apiVersion: velero.io/v1
kind: Backup
metadata:
  name: <backup-name>
  namespace: <protected-ns>
spec:
  includedNamespaces:
  - <app-ns>
  storageLocation: velero-sample-1
----

.. Monitor the progress of the `VolumeSnapshotBackup` CRs by doing the following:
... To check the progress of all the `VolumeSnapshotBackup` CRs, run the following command:
+
[source, terminal]
----
$ oc get vsb -n <app-ns>
----

... To check the progress of a specific `VolumeSnapshotBackup` CR, run the following command:
+
[source,terminal]
----
$ oc get vsb <vsb-name> -n <app-ns> -ojsonpath="{.status.phase}`
----

.. Wait several minutes until the `VolumeSnapshotBackup` CR has the status `Completed`.
.. Verify that there is at least one snapshot in the object store that is given in the Restic `secret`. You can check for this snapshot in your targeted `BackupStorageLocation` storage provider that has a prefix of `/<OADP-namespace>`.

. To restore data, complete the following steps:

> 

Author: @akarol @tareqalayan ASAP I need someone in QE to verify that lines 129-143 below is a correct way to
"Make sure the application namespace is deleted, as well as any volumeSnapshotContents that were created during backup."  David Vaanunu is the guy who knows this best, but I can't tag him here (will try in Jira)
